### PR TITLE
add non-maximum suppression (NMS) function to torchsharp

### PR DESCRIPTION
**Pull request by huangyangyu**
This is a C# implementation of [NMS](https://pytorch.org/vision/0.9/ops.html?highlight=nms#torchvision.ops.nms), which is verified in object detection feature.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=720) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=720) before we can accept your contribution.<br/><br/>:x: huangyangyu [sign now](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=720)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/TorchSharp?pullRequest=720) it.</sub>

File: src/TorchSharp/TorchVision/Ops.cs [hunk 1]
```diff
@@ -30,5 +33,48 @@ public static Tensor sigmoid_focal_loss(Tensor inputs, Tensor targets, float alp
                 return loss.MoveToOuterDisposeScope();
             }
         }
+
+        public static Tensor nme(Tensor boxes, Tensor scores, double iou_threshold = 0.5)
```
Comment (NiklasGustafsson) [inline src/TorchSharp/TorchVision/Ops.cs:37]: Is it NMS or NME?

File: src/TorchSharp/TorchVision/Ops.cs [hunk 2]
```diff
@@ -30,5 +33,48 @@ public static Tensor sigmoid_focal_loss(Tensor inputs, Tensor targets, float alp
...
+        {
+            using (var nmsScope = torch.NewDisposeScope()) {
+                var x1 = boxes[.., 0];
```
Comment (NiklasGustafsson) [inline src/TorchSharp/TorchVision/Ops.cs:40]: System.Range is only available in later version of .NET, and we need to be compatible with .NET FX 4.7.2. Therefore, you need to use the uglier syntax that System.Range turns into.

File: src/TorchSharp/TorchVision/Ops.cs [hunk 3]
```diff
@@ -30,5 +33,48 @@ public static Tensor sigmoid_focal_loss(Tensor inputs, Tensor targets, float alp
...
+                    if (order.numel() == 1) {
+                        i = order.item<long>();
+                        keep.Add(i);
+                        break;
+                    } else {
+                        i = order[0].item<long>();
+                        keep.Add(i);
+                    }
+
+                    var xx1 = x1[order[1..]].clamp(min: x1[i]);
```
Comment (NiklasGustafsson) [inline src/TorchSharp/TorchVision/Ops.cs:60]: Same here -- can't use ranges in implementation code.

File: src/TorchSharp/TorchVision/Ops.cs [hunk 4]
```diff
@@ -1,6 +1,9 @@
 // Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
 
 using static TorchSharp.torch;
+using System.Range;
```
Comment (NiklasGustafsson) [inline src/TorchSharp/TorchVision/Ops.cs:4]: Not available on all versions of .NET that we are targeting.

File: src/TorchSharp/TorchVision/Ops.cs [hunk 5]
```diff
@@ -30,5 +33,48 @@ public static Tensor sigmoid_focal_loss(Tensor inputs, Tensor targets, float alp
                 return loss.MoveToOuterDisposeScope();
             }
         }
+
```
Comment (NiklasGustafsson) [inline src/TorchSharp/TorchVision/Ops.cs:36]: Please add doc comments.

**Review by NiklasGustafsson — CHANGES_REQUESTED** [review]

**Review by NiklasGustafsson — COMMENTED** [review]
Also, please add some unit tests.

File: src/TorchSharp/TorchVision/Ops.cs (hunk 1 above)
Comment (huangyangyu) [reply src/TorchSharp/TorchVision/Ops.cs]: Thanks for pointing out the typo.

File: src/TorchSharp/TorchVision/Ops.cs (hunk 2 above)
Comment (huangyangyu) [reply src/TorchSharp/TorchVision/Ops.cs]: changed it to tensor.select(..)

File: src/TorchSharp/TorchVision/Ops.cs (hunk 3 above)
Comment (huangyangyu) [reply src/TorchSharp/TorchVision/Ops.cs]: changed it to tensor indices, like this: torch.arange(..).

File: src/TorchSharp/TorchVision/Ops.cs (hunk 4 above)
Comment (huangyangyu) [reply src/TorchSharp/TorchVision/Ops.cs]: Removed the unsupported dependencies.

File: src/TorchSharp/TorchVision/Ops.cs (hunk 5 above)
Comment (huangyangyu) [reply src/TorchSharp/TorchVision/Ops.cs:37]: Added.

File: src/TorchSharp/TorchVision/Ops.cs (hunk 2 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/TorchVision/Ops.cs]: You could also use: 

```C#
var idx = TensorIndex.Slice(); // Default values is the same as '..'
var x1 = boxes[idx, 0];
```

I don't know if that or 'select()' is faster.

File: src/TorchSharp/TorchVision/Ops.cs (hunk 3 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/TorchVision/Ops.cs]: TensorIndex.Slice(1,null) may be faster.

**Review by NiklasGustafsson — APPROVED** [review]

**Pull request merged**